Ports/generate-available-ports.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+
+import os
+import re
+import sys
+import asyncio
+
+CONCURRENCY = 20
+GIT_HASH_REGEX = re.compile(r'^[0-9a-f]{40}$')
+
+
+def escape_markdown(value):
+    return value.replace("_", "\\_").replace("*", "\\*")
+
+
+def format_version(version):
+    return '' if version == "git" \
+        else version[0:7] if GIT_HASH_REGEX.match(version) \
+        else escape_markdown(version)
+
+
+HEADERS = {
+    'port': {'caption': 'Port', 'min_width': 51, 'formatter': lambda name: f'[`{name}`]({name}/)'},
+    'description': {'caption': 'Name', 'min_width': 63, 'formatter': escape_markdown},
+    'version': {'caption': 'Version', 'min_width': 24, 'formatter': format_version},
+    'website': {'caption': 'Website', 'min_width': 78}
+}
+
+PROPERTIES = HEADERS.keys()
+
+
+def die(message, code):
+    sys.stderr.write(f"\n{message}\n\n")
+    sys.exit(code)
+
+
+def join(data, separator, prefix='', postfix=''):
+    return f"{prefix}{separator.join(data)}{postfix}"
+
+
+async def read_port_properties(port):
+    if not os.path.exists(port + '/package.sh'):
```

Comment:
A few comments on the python here:

- we should prefer to use pathlib over os.path <https://docs.python.org/3/library/pathlib.html>
- Can we not do `f'{port}/package.sh showproperty ....'` rather than doing a cd + exec?
- What does calling `sys.stderr.write/flush` from each async task do to our throughput in reading all these files at once?
- I'm not convinced that repeated calls to `sys.<stream>.write` in a row is particularly pythonic. Is there a way we can create strings and then write them all at once? Mostly thinking about the final writeout of the entire file.
- Why are we writing directly to stdout instead of passing a file argument in?

Reply:
thanks! answering to each point:
* got rusty on that; did not even it was a thing—definitely it is—will switch to it; thanks!
* it is needed in some form (`package.sh` contains relative call in the prologue), but can be simplified;
* itʼs needed to see the progress; on machine the penalty is negligible (total time sticks around 3.14s).
* i was not aiming to be 100% pythonic here; iʼll make it less offensive by concatenating it. 😉
* it was just easy (not dealing with opening files etc.); shall i take it as a request for change? 🙂